### ISSUE TITLE
False positive for `no-unused-services` when component is passed to function expressions

### ISSUE THREAD

**Issue by tehhowch** [ember-cli/eslint-plugin-ember#1258]
While evaluating this rule on apps at work, I discovered it triggered in the following scenario:

```js
import Component from '@glimmer/component';
import { inject as service } from '@ember/service';

const SomeUtilityFn = (component) => component.someService.someMethod();

export default class MyComponent extends Component {
  @service someService;

  get foo() {
    return SomeUtilityFn(this);
  }
}
```

Given that the helper method is defined in this file (vs imported), I didn't expect the rule to be triggered.

**Comment by bmish** [comment]
This is a good edge case that we could try fixing. It might be tricky to statically detect this usage of the service, although it should be possible.

I'm not sure if this is the most elegant way, but something like this:

1. Inside the component, check for any function calls with `this` as an argument.
2. For any such functions, check if they are defined in the same file.
3. If they are, check inside their function scope to see if the relevant parameter has a service accessed off of it.

**Comment by tehhowch** [comment]
The same logic would also improve the require-computed-property-dependencies rule, which fails to deduce a missing declared dependency for the same pattern:

```js
import Component from '@ember/component';
import { computed } from '@ember/object';

const SomeUtilityFn = (component) => component.prop;

export default Component.extend({
  prop: 1,
  bar: 2,

  foo: computed('bar', function () {
    return bar && SomeUtilityFn(this);
  }), // require-computed-property-dependencies fails to warn about undeclared dependency on `prop`
});
```
